xbmc/windowing/gbm/GBMUtils.cpp

```diff
@@ -0,0 +1,555 @@
+                             DRM_MODE_PAGE_FLIP_EVENT,
+                             &flip_happening);
+
+  if(ret)
+  {
+    CLog::Log(LOGDEBUG, "CGBMUtils::%s - failed to queue DRM page flip", __FUNCTION__);
+    return false;
+  }
+
+  return true;
+}
+
+void CGBMUtils::FlipPage()
+{
+  if(WaitingForFlip())
+  {
+    return;
+  }
+
+  flip_happening = QueueFlip();
+
+  if(g_Windowing.NoOfBuffers() >= 3 && gbm_surface_has_free_buffers(m_drm->gbm->surface))
+  {
+    return;
+  }
+
+  WaitingForFlip();
+}
+
+bool CGBMUtils::GetResources()
+{
+  m_drm_resources = drmModeGetResources(m_drm->fd);
+  if(!m_drm_resources)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool CGBMUtils::GetConnector()
+{
+  for(auto i = 0; i < m_drm_resources->count_connectors; i++)
+  {
+    m_drm_connector = drmModeGetConnector(m_drm->fd,
+                                          m_drm_resources->connectors[i]);
+    if(m_drm_connector->connection == DRM_MODE_CONNECTED)
+    {
+      CLog::Log(LOGDEBUG, "CGBMUtils::%s - found connector: %d", __FUNCTION__,
+                                                                 m_drm_connector->connector_type);
+      break;
+    }
+    drmModeFreeConnector(m_drm_connector);
+    m_drm_connector = nullptr;
+  }
+
+  if(!m_drm_connector)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool CGBMUtils::GetEncoder()
+{
+  for(auto i = 0; i < m_drm_resources->count_encoders; i++)
+  {
+    m_drm_encoder = drmModeGetEncoder(m_drm->fd, m_drm_resources->encoders[i]);
+    if(m_drm_encoder->encoder_id == m_drm_connector->encoder_id)
+    {
+      CLog::Log(LOGDEBUG, "CGBMUtils::%s - found encoder: %d", __FUNCTION__,
+                                                               m_drm_encoder->encoder_type);
+      break;
+    }
+    drmModeFreeEncoder(m_drm_encoder);
+    m_drm_encoder = nullptr;
+  }
+
+  if(!m_drm_encoder)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool CGBMUtils::GetPreferredMode()
+{
+  for(auto i = 0, area = 0; i < m_drm_connector->count_modes; i++)
+  {
+    drmModeModeInfo *current_mode = &m_drm_connector->modes[i];
+
+    if(current_mode->type & DRM_MODE_TYPE_PREFERRED)
+    {
+      m_drm->mode = current_mode;
+      CLog::Log(LOGDEBUG,
+                "CGBMUtils::%s - found preferred mode: %dx%d%s @ %d Hz",
+                __FUNCTION__,
+                m_drm->mode->hdisplay,
+                m_drm->mode->vdisplay,
+                m_drm->mode->flags & DRM_MODE_FLAG_INTERLACE ? "i" : "",
+                m_drm->mode->vrefresh);
+      break;
+    }
+
+    auto current_area = current_mode->hdisplay * current_mode->vdisplay;
+    if (current_area > area)
+    {
+      m_drm->mode = current_mode;
+      area = current_area;
+    }
+  }
+
+  if(!m_drm->mode)
+  {
+    CLog::Log(LOGDEBUG, "CGBMUtils::%s - failed to find preferred mode", __FUNCTION__);
+    return false;
+  }
+
+  return true;
+}
+
+drm * CGBMUtils::InitDrm()
+{
+  const char *device = "/dev/dri/card0";
```

Comment:
You'll want to stick this in a loop, iterating all the cards until one returns you a resources pointer with at least one CRTC/encoder/connector. Otherwise you can load virtual devices (like VGEM), or inert rendering-only devices (a discrete GPU on a multi-GPU system).